In [697]:
sum1 = 3
sum2 = 4
name = "frank"
print("everything should be",sum1,sum2,name * 2)

everything should be 3 4 frankfrank


In [698]:
my_list = ["Frank",5,"he's very rich", True]
print(my_list)

['Frank', 5, "he's very rich", True]


In [699]:
empty_list = []
empty_list.append("knocking on the door")
print(empty_list)

['knocking on the door']


In [700]:
combine = my_list + empty_list
print(combine)

['Frank', 5, "he's very rich", True, 'knocking on the door']


In [701]:
test_list = [1,2,3,2,4]
test_list.remove(2)
print(test_list)
print( len(test_list))
print( max(test_list))
print( min(test_list))
print( sum(test_list))



[1, 3, 2, 4]
4
4
1
10


In [702]:
"he's very rich" in my_list
test_list.count(4)

test_list.sort()

print("sorted list",(test_list))

test_list.reverse()
print("reverse list",(test_list))


sorted list [1, 2, 3, 4]
reverse list [4, 3, 2, 1]


In [703]:
another_list = ["hello", "my","dearest", "loving", "wife"]
print(another_list[-3])

dearest


In [704]:
nested_list = [[2,3,6,7],[9,7,4,8],[3,2,1]]
print(nested_list[1][2])

my_slice= nested_list[-2:]
print(my_slice)

nested_list[1] = "change"
print(nested_list[1])

del(nested_list[2][2])
print(nested_list[2][1])

next_item = nested_list.pop()
print(next_item)




4
[[9, 7, 4, 8], [3, 2, 1]]
change
2
[3, 2]


In [705]:
list1 = [1,2,3]
list2 = list1.copy()

list1.append(4)
print("list1:",list1)
print("list2:",list2)

list1: [1, 2, 3, 4]
list2: [1, 2, 3]


In [706]:
import copy

list1 = [1,2,3]
list2 = ["list withing a list", list1]
list3 = copy.deepcopy(list2)

print("before appending to list:")
print("list2:", list2)
print("list3:", list3, "\n")

list1.append(4)
print("after appending to list:")
print("list2:", list2)
print("list3:", list3)


before appending to list:
list2: ['list withing a list', [1, 2, 3]]
list3: ['list withing a list', [1, 2, 3]] 

after appending to list:
list2: ['list withing a list', [1, 2, 3, 4]]
list3: ['list withing a list', [1, 2, 3]]


Tuples and string


In [707]:
mytuple = (1,2,3)
print(mytuple)

(1, 2, 3)


In [708]:
my_list = [2,3,1,4,4]
another_tuple = tuple(my_list)
another_tuple

(2, 3, 1, 4, 4)

Operations performed on list can be done also in tuples

In [709]:
another_tuple.count(4)

2

In [710]:
my_string = "hello world"
my_string.replace("o",
                  "a")

'hella warld'

In [711]:
" ".join(["i","can","see", "you"])

'i can see you'

In [712]:
name = "Frank"
age = 10
city = "Lagos"
f"My name is {name} i am {age} and i live in {city}"

'My name is Frank i am 10 and i live in Lagos'

Dictionary and sets

In [713]:
my_dict = {"name":"Frank","age": 10,"city":"Lagos"}
print(my_dict)
my_dict["name"]
my_dict["Sex"] = "Male"
print(my_dict)
len(my_dict)
"name" in my_dict
my_dict.keys()
my_dict.values()
"Male"in my_dict.values()
my_dict.items()

{'name': 'Frank', 'age': 10, 'city': 'Lagos'}
{'name': 'Frank', 'age': 10, 'city': 'Lagos', 'Sex': 'Male'}


dict_items([('name', 'Frank'), ('age', 10), ('city', 'Lagos'), ('Sex', 'Male')])

First table

name  age  city
Frank  30  lagos
Bob    20  Abeokuta
Harry  15  Ibadan

storing this in a dictionary

In [714]:
my_table_dict = {"name": ["Frank","Bob","Harry"],
                 "Age":[30,20.15],
                 "City":["Lagos","Abeokuta","Ibadan"]}
print (my_table_dict)
type(my_table_dict)


{'name': ['Frank', 'Bob', 'Harry'], 'Age': [30, 20.15], 'City': ['Lagos', 'Abeokuta', 'Ibadan']}


dict

SETS are unordered, mutable collections of immutable objects that cannot contain duplicates.

In [715]:
my_set = {1,2,3,4,5,6,7}
type(my_set)
my_set.add(8)
my_set.remove(8)
my_set

{1, 2, 3, 4, 5, 6, 7}

sets can also check len min max or relationship with (in)

the main purpose of set is to perform sets operation that compare or combine different sets.
python sets supports many common mathematical set operation like union,intersection,difference and checking whether one set is  a subset of another.

In [716]:
set1 = {1,3,5,6}
set2 = {1,2,3,4}
set1.union(set2)
set1.intersection(set2)
set1.difference(set2)
set2.difference(set1)
set2.issubset(set1)



False

turning list intoa set

In [717]:
my_list = [1,2,2,2,3,3,4,5,5,5,6]
list(set(my_list))

[1, 2, 3, 4, 5, 6]

Numpy arrays.



In [718]:
import numpy as np
import pandas as pd
import matplotlib as matplotlib


It defines a new series by passing a collection of homogeneous data like ndarraysor list, along with a list of associate indexes to pd.series():

In [719]:
my_series = pd.Series( data = [2,3,5,4],        
                      index = ['a','b','c','d'])
my_series

a    2
b    3
c    5
d    4
dtype: int64

We can also create a series from a dictionary, in which case the dictionary keys act as the labels and the values act as the data

In [720]:
my_dict = {'x': 2,'a': 5, 'b': 4, 'c': 8}
my_series2 = pd.Series(my_dict)
my_series2

x    2
a    5
b    4
c    8
dtype: int64

In [721]:
my_series['a']

2

slice it


In [722]:
my_series[1:3]


b    3
c    5
dtype: int64

checking for mean in numpy

In [723]:
np.min(my_series)

2

DataFrame Creation and Indexing

A DataFrame is a 2D table with labeled columns that can each hold different typesof data
DataFrames are essentially a python implementation of the types of table you'dseei Excell workbook or  SQL 
DataFrames are  the defacto standard data structure.

we can create a DataFrame out of a variety of data sources like dictionaries, 2D and series using the pd.DataFrame() function. 

In [724]:
# create a dictionary with some different data types as values

my_dict = {"name": ["Joe","Mary","Frans"],     # List
           "age": np.array([10,15,20]),        # numpy array
           "weight": (75,123,239),             # tuple
           "height": pd.Series([4.5,5,6.1],    # pd.Series
                               index = [0,1,2,]), # List
           "siblings": [1,3,4],
           "gender": ["M","F","M"]}
df = pd.DataFrame(my_dict)                      # converting the dictionary to DataFrame
df                                              #show the dataframe

name  age  weight  height  siblings gender
0    Joe   10      75     4.5         1      M
1   Mary   15     123     5.0         3      F
2  Frans   20     239     6.1         4      M

In [725]:
my_dict2 = {"name": ["Joe","Mary","Frans"],     # List
           "age": np.array([10,15,20]),        # numpy array
           "weight": (75,123,239),             # tuple
           "height": pd.Series([4.5,5,6.1]),    # pd.Series
                               #index = [0,1,2,]), # List
           "siblings": [1,3,4],
           "gender": ["M","F","M"]}
df = pd.DataFrame(my_dict)                      # converting the dictionary to DataFrame
df             

name  age  weight  height  siblings gender
0    Joe   10      75     4.5         1      M
1   Mary   15     123     5.0         3      F
2  Frans   20     239     6.1         4      M

In [726]:
# Providing a custom row labels when creating a DataFrame by adding the index arguement

df2 = pd.DataFrame(my_dict2,
                   index = my_dict["name"])
df2

name  age  weight  height  siblings gender
Joe      Joe   10      75     NaN         1      M
Mary    Mary   15     123     NaN         3      F
Frans  Frans   20     239     NaN         4      M

In [727]:
# We Can also Add or delete columns in a DataFrame the same way we would when dealing with a dictionary

df2["weight"]
df2.age 

Joe      10
Mary     15
Frans    20
Name: age, dtype: int32

In [728]:
# to delete a column
del df2['name']
df2

age  weight  height  siblings gender
Joe     10      75     NaN         1      M
Mary    15     123     NaN         3      F
Frans   20     239     NaN         4      M

In [729]:
# Add a column
df2["IQ"] = [130, 105, 115]
df2

age  weight  height  siblings gender   IQ
Joe     10      75     NaN         1      M  130
Mary    15     123     NaN         3      F  105
Frans   20     239     NaN         4      M  115

In [730]:
# we can also select both rows and columns by label with df.loc[row,column]

df2.loc["Joe"]

age          10
weight       75
height      NaN
siblings      1
gender        M
IQ          130
Name: Joe, dtype: object

In [731]:
df2.loc["Joe","IQ"]

130

In [732]:
df2.iloc[0]

age          10
weight       75
height      NaN
siblings      1
gender        M
IQ          130
Name: Joe, dtype: object

In [733]:
df2.iloc[0, 5]

130

In [734]:
df2.iloc[0:2, 5:8]

IQ
Joe   130
Mary  105

In [735]:
boolean_index = [False, True, True]
df2[boolean_index] 

age  weight  height  siblings gender   IQ
Mary    15     123     NaN         3      F  105
Frans   20     239     NaN         4      M  115

In [736]:
# Creating a boolean sequence with a logical comparison

boolean_index = df2["age"] > 12
# use the index to get the rows where age > 12

df2[boolean_index] 

age  weight  height  siblings gender   IQ
Mary    15     123     NaN         3      F  105
Frans   20     239     NaN         4      M  115

In [737]:
df2[ df2["age"] > 12 ]

age  weight  height  siblings gender   IQ
Mary    15     123     NaN         3      F  105
Frans   20     239     NaN         4      M  115

Exploring DataFrames

Exploring data is an important first step in most data analysis.
DataFrames contains varieties of function to help you and summarize
the data they contain.

In [738]:
titanic_train = pd.read_csv("./train.csv")
type(titanic_train)

pandas.core.frame.DataFrame

In [739]:
titanic_train.shape     # check dimensions

(891, 12)

The output shows that Titanic train data has 891 rows and 12 columns
we can check the first n rows of the data the df.head() 

In [740]:
titanic_train.head(6)      # check the first 6 rows 

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q

We can equally check last few rows with df.tail(); 

In [741]:
titanic_train.tail(6)

PassengerId  Survived  Pclass                                      Name  \
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket    Fare Cabin Embarked  
885  female  39.0      0      5      382652  29.125   NaN        Q  
886    male  27.0      0      0      211536  13.000   NaN        S  
887  female  19.0      0      0      112053  30.000   B42        S  
888  female   NaN      1      2  W./C. 6607  23.450   NaN        S  
889    male  26.0      0      0      111369  30.000  C148        C  
890    male  32.0      0      0      370376   7.750   NaN        Q

In [742]:
# We can access and assign new row indexes with df.index:

titanic_train.index = titanic_train["Name"]   # Set index to name
del titanic_train["Name"]                     # Delete name column
print(titanic_train.index[0:10])

Index(['Braund, Mr. Owen Harris',
       'Cumings, Mrs. John Bradley (Florence Briggs Thayer)',
       'Heikkinen, Miss. Laina',
       'Futrelle, Mrs. Jacques Heath (Lily May Peel)',
       'Allen, Mr. William Henry', 'Moran, Mr. James',
       'McCarthy, Mr. Timothy J', 'Palsson, Master. Gosta Leonard',
       'Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)',
       'Nasser, Mrs. Nicholas (Adele Achem)'],
      dtype='object', name='Name')


In [743]:
# To Access the column labels with df.columns

titanic_train.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

To get a quick statistical summary of your data set. 
The summary includes the Mean, Median,min,max and few key percentiles  for 
numeric columns.


In [744]:
titanic_train.describe()     # Summarize the first 6 columns

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [750]:
titanic_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 891 entries, Braund, Mr. Owen Harris to Dooley, Mr. Patrick
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Sex          891 non-null    object 
 4   Age          714 non-null    float64
 5   SibSp        891 non-null    int64  
 6   Parch        891 non-null    int64  
 7   Ticket       891 non-null    object 
 8   Fare         891 non-null    float64
 9   Cabin        204 non-null    object 
 10  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(4)
memory usage: 83.5+ KB
